http-server/datahike/http/server.clj

```diff
@@ -0,0 +1,172 @@
+(ns datahike.http.server
+  (:gen-class)
+  (:refer-clojure :exclude [read-string filter])
+  (:require
+   [clojure.string :as str]
+   [clojure.edn :as edn]
+   [datahike.api.specification :refer [api-specification ->url]]
+   [datahike.http.middleware :as middleware]
+   [datahike.readers :refer [edn-readers]]
+   [datahike.transit :refer [read-handlers write-handlers]]
+   [datahike.api :refer :all :as api]
+   [reitit.ring :as ring]
+   [reitit.coercion.spec]
+   [reitit.swagger :as swagger]
+   [reitit.swagger-ui :as swagger-ui]
+   [reitit.ring.coercion :as coercion]
+   [reitit.ring.middleware.muuntaja :as muuntaja]
+   [reitit.ring.middleware.multipart :as multipart]
+   [reitit.ring.middleware.parameters :as parameters]
+   [ring.middleware.cors :refer [wrap-cors]]
+   [muuntaja.core :as m]
+   [clojure.spec.alpha :as s]
+   [datahike.spec :as spec]
+   [datahike.impl.entity :as de]
+   [taoensso.timbre :as log]
+   [ring.adapter.jetty :refer [run-jetty]]
+   [spec-tools.core :as st])
+  (:import [datahike.datom Datom]))
+
+(defn generic-handler [f edn-args edn-ret]
+  (fn [request]
+    (let [{{body :body} :parameters
+           :keys [headers]} request
+          _ (log/trace "req-body" f body (= f #'api/create-database))
+          ;; TODO move this to client
+          ret-body
+          (cond (= f #'api/create-database)
+                ;; remove remote-peer and re-add
+                (assoc
+                 (apply f (dissoc (first body) :remote-peer) (rest body))
+                 :remote-peer (:remote-peer (first body)))
+
+                (= f #'api/delete-database)
+                (apply f (dissoc (first body) :remote-peer) (rest body))
+
+                :else
+                (apply f body))]
+      (log/trace "ret-body" ret-body)
+      {:status 200
+       :body
+       (when-not (headers "no-return-value")
+         ret-body)})))
+
+(declare create-routes)
+
+(defn extract-first-sentence [doc]
+  (str (first (str/split doc #"\.\s")) "."))
+
+;; This code expands and evals the server route construction given the
+;; API specification. 
+;; TODO This would not need macro-expansion if s/spec would not be a macro.
+(eval
+ `(defn ~'create-routes [~'config]
+    ~(vec
+      (for [[n {:keys [args doc edn-args edn-ret supports-remote?]}] api-specification
+            :when                                                    supports-remote?]
+        `[~(str "/" (->url n))
+          {:swagger {:tags ["API"]}
+           :post    {:operationId ~(str n)
```

Comment:
The current clients are already broken, because now query works differently. I agree that breaking them is not good though. 

My main goal is to have a general (regular) interface that can be expanded like this, which required effectively passing through all arguments to each function (endpoint) as a blob in the POST body, as this works in general (and in my opinion is how a RPC mechanism should be designed). My understanding from a brief web search is that GET is now also allowed to have a body, which means we could declare which functions do state changes to a DB in the specification and then pick POST for `transact` and `load-entities` only.  The main reason to map everything that does not change the state to GET is for HTTP caching. For this to work the resource identification would have to include all arguments that are in the POST body now though, and squeezing them into the URL (or headers?) can be very tricky because of all the encoding standards and potential field limits HTTP has. Maybe there is a simple way to do it that I don't know though. How would you pass the `args` to GET?

Reply:
A quick test to replace POST by GET did not work though.

Reply:
If it's all POST or all GET I would prefer it all POST. q and transact were both POST requests in the datahike-server as well. There are some queries using get in datahike-server. I don't think it's really important if it breaks implementations anyway.